### PR TITLE
datacollector: store separate snapshots of model data per step

### LINKED ISSUE #2128: Issue with data collection of PropertyLayer objects
**Describe the bug**
When using a PropertyLayer class with DataCollector, the data is not captured properly at each model step.  Specifically, for each model step, the grid of values for the PropertyLayer instance are stored only as they appear in the final model step, i.e. all steps store the same array of values.

**Expected behavior**
The DataCollector should store at each step the value of the PropertyLayer at that step as a model variable.  This would allow visualization of the distribution of the variable at each step of the model to examine, for example, the accumulation of pollution in the Sugarscape model.

**To Reproduce**
The toy model below reproduces the issue.  Check the 'Pollution' column of the model_results object to see the problem.  Minimally, one would expect a grid of 0's for the first model step, which is not the case here either.

```
import mesa

class testAgent(mesa.Agent):
    def __init__(self, unique_id, model, pos):
        super().__init__(unique_id, model)
        self.pos = pos
    def pollute(self):
        self.model.grid.properties["pollution"].modify_cell(self.pos, lambda p: p+1)

class testModel(mesa.Model):
    def __init__(self, width = 20, height = 20, initial_population=200):
        super().__init__()
        self.width = width
        self.height = height
        self.initial_population = initial_population
        
        # initiate activation schedule
        self.schedule = mesa.time.RandomActivationByType(self) #activate different type of agents in different order each time
        # initiate mesa grid class
        self.grid = mesa.space.MultiGrid(self.width, self.height, torus=False)
        #add pollution property layer
        pollution_layer = mesa.space.PropertyLayer(name = "pollution", width = self.width, height = self.height, default_value = 0)
        self.grid.add_property_layer(pollution_layer)
        
        self.datacollector = mesa.DataCollector(model_reporters = {"Pollution": lambda m: m.grid.properties["pollution"].data})
        
        agent_id = 0
        for i in range(self.initial_population):
            # get agent position
            x = self.random.randrange(self.width)
            y = self.random.randrange(self.height)
            agent = testAgent(agent_id, self, (x, y))
            # place agent
            self.grid.place_agent(agent, (x, y))
            self.schedule.add(agent)
            agent_id += 1
    
    def randomizeAgents(self):
        agents_shuffle = list(self.schedule.agents_by_type[testAgent].values())
        self.random.shuffle(agents_shuffle)

        return agents_shuffle
    
    def step(self):
        agents_shuffle = self.randomizeAgents()
        for agent in agents_shuffle:
            agent.pollute()
        
        self.datacollector.collect(self)
       
    def run_model(self, step_count=20):
        for i in range(step_count):
            self.step()
            
model = testModel()
model.run_model()
model_results = model.datacollector.get_model_vars_dataframe()
```

**Additional context**
This should be running on the most recent version of Mesa (my version was updated about a week ago).

### PR DESCRIPTION
**Pull request by EwoutH**
This pull request addresses issue #2128, which involves incorrect data collection for `PropertyLayer` objects when using Mesa's `DataCollector` module. Previously, the `DataCollector` was maintaining references to the `PropertyLayer` data array, leading to every step showing identical data due to changes propagating backward.

**Changes Introduced**:
- Updated the `collect` method of the `DataCollector` to use `deepcopy` for model-level data collection.
- Ensures that unique snapshots of data are collected at each model step rather than maintaining references.

Resolves #2128.

### CODE REVIEW

**Comment by EwoutH** [comment]
@dylan-munson this PR should resolve your issue (I tested it locally and for me it does). You can install this branch with:
```bash
pip install -U -e git+https://github.com/projectmesa/mesa@datacollector_deepcopy#egg=mesa
```
Thanks for reporting it and please let us know if this fixed it for you.

**Comment by dylan-munson** [comment]
@EwoutH Thanks for looking into this; I will try this in a bit.  Just to check, I will be able to return to the main branch of the mesa repo by just running a usual pip install update right?

**Comment by dylan-munson** [comment]
I can confirm that after installing this dev branch the issue has been fixed.  Thank you for the help!  I'll keep my eyes open for when this makes it into the main branch.

File: mesa/datacollection.py [hunk 1]
```diff
@@ -197,17 +198,19 @@ def collect(self, model):
             for var, reporter in self.model_reporters.items():
                 # Check if lambda or partial function
                 if isinstance(reporter, (types.LambdaType, partial)):
-                    self.model_vars[var].append(reporter(model))
+                    self.model_vars[var].append(deepcopy(reporter(model)))
```
Comment (rht) [inline mesa/datacollection.py:201]: Needs a brief explanatory comment before this line. Otherwise someone might mistakenly remove the deepcopy.

Comment (EwoutH) [reply mesa/datacollection.py:203]: Agreed, done.

**Comment by EwoutH** [comment]
@dylan-munson Thanks for validating that this fix resolves your issue!

@rht Thanks for reviewing. I added the comment and resolved the merge conflict. Please squash when merging.

**Review by rht — APPROVED** [review]

**Comment by rht** [comment]
Can't merge due to conflicts.

**Comment by rht** [comment]
Squash and merge worked.

**Comment by EwoutH** [comment]
@dylan-munson I cherry picked this PR to the 2.3 maintenance branch, and tagged Mesa 2.3.1 which includes this fix. So you can now also use that version.

**Pull request merged**